bait-files/src/test/java/com/google/sps/LoginServletTest.java

```diff
@@ -0,0 +1,85 @@
+package com.google.sps;
+  
+import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.google.cloud.datastore.testing.LocalDatastoreHelper;
+import com.google.sps.servlets.User;
+import com.google.sps.servlets.LoginServlet;
+import com.googlecode.objectify.ObjectifyFactory;
+import com.googlecode.objectify.ObjectifyService;
+import com.googlecode.objectify.util.Closeable;
+import com.google.sps.servlets.UserAccessor;
+import java.io.IOException;
+import java.util.List;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.concurrent.TimeoutException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.AfterClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class LoginServletTest {
+  static LocalDatastoreHelper datastoreHelper = LocalDatastoreHelper.create(1.0);
+
+  private Closeable objectify;
+
+  @BeforeClass
+  public static void oneTimeSetUp() throws InterruptedException, IOException, TimeoutException {
+      datastoreHelper.start();
+  }
+
+  @Before 
+  public void setUp() throws InterruptedException, IOException {
+    datastoreHelper.reset();
+    ObjectifyFactory factory =
+        new ObjectifyFactory(datastoreHelper.getOptions().getService());
+    ObjectifyService.init(factory);
+    ObjectifyService.register(User.class);
+    objectify = ObjectifyService.begin();
+  }
+
+  @After
+  public void tearDown() throws InterruptedException, IOException, TimeoutException {
+        objectify.close();
+  }
+
+  @AfterClass 
+    public static void oneTimeTearDown() throws InterruptedException, IOException, TimeoutException {
+      datastoreHelper.stop();
+  }
+
+  @Test
+  public void doPost_newUser_oneUser() throws IOException, ServletException {
+    // add a User object
+
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    HttpServletResponse response = mock(HttpServletResponse.class);
+    UserAccessor userAccessor = mock(UserAccessor.class);
+
+    when(userAccessor.getUserId()).thenReturn("123");
+    when(request.getParameter("username")).thenReturn("Drew");
+    when(request.getParameter("first-name")).thenReturn("Andrew");
+    when(request.getParameter("last-name")).thenReturn("Seaman");
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter printWriter = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(printWriter);
```

Comment:
Out of curiosity... maybe see if you could get away with deleting this stuff... in your implementation for doPost for LoginServlet, I don't think you're actually calling getWriter() or doing anything with the writer, so mocking it might be a waste.  The only thing you're doing with the response in doPost is calling sendRedirect, so I guess it depends if that method internally uses the writer...

In any case, I'd say try deleting these lines and see if the test still passes.  If not, just add them back in =D

Reply:
Removed